vyper/codegen/stmt.py

```diff
@@ -254,37 +254,43 @@ def _parse_For_range(self):
 
         # Get arg0
         arg0 = self.stmt.iter.args[0]
-        num_of_args = len(self.stmt.iter.args)
+        arg1 = self.stmt.iter.args[1] if len(self.stmt.iter.args) > 1 else None
 
         kwargs = {
             s.arg: Expr.parse_value_expr(s.value, self.context)
             for s in self.stmt.iter.keywords or []
         }
 
         # Type 1 for, e.g. for i in range(10): ...
-        if num_of_args == 1:
+        if arg1 is None:
             n = Expr.parse_value_expr(arg0, self.context)
             start = IRnode.from_list(0, typ=iter_typ)
             rounds = n
             rounds_bound = kwargs.get("bound", rounds)
 
         # Type 2 for, e.g. for i in range(100, 110): ...
-        elif self._check_valid_range_constant(self.stmt.iter.args[1]).is_literal:
+        elif self._check_valid_range_constant(arg1).is_literal:
             arg0_val = self._get_range_const_value(arg0)
-            arg1_val = self._get_range_const_value(self.stmt.iter.args[1])
+            arg1_val = self._get_range_const_value(arg1)
             start = IRnode.from_list(arg0_val, typ=iter_typ)
             rounds = IRnode.from_list(arg1_val - arg0_val, typ=iter_typ)
             rounds_bound = rounds
 
         # Type 3 for, e.g. for i in range(x, x + 10): ...
-        else:
-            arg1 = self.stmt.iter.args[1]
+        elif "bound" not in kwargs:
             rounds = self._get_range_const_value(arg1.right)
             start = Expr.parse_value_expr(arg0, self.context)
             _, hi = start.typ.int_bounds
             start = clamp("le", start, hi + 1 - rounds)
             rounds_bound = rounds
 
+        # Type 4 for, e.g. for i in range(x, y, bound=N): ...
+        else:
+            start = Expr.parse_value_expr(arg0, self.context)
+            end = Expr.parse_value_expr(arg1, self.context)
+            rounds = IRnode.from_list(["sub", end, clamp("le", start, end)])
```

Comment:
for re-used variables we currently need to use `with end.cache_with_complex("end") as (b1, end):` since that handles side effects

Reply:
you can see this with something like
```
@internal
def some_side_effects() -> uint256:  # for illustration
    self.counter += 1
    return 10

@external
def foo():
    for i in range(1, self.some_side_effects(), bound=100):
        ...
    assert self.counter == 1
```

Reply:
This example raises `StateAccessViolation: May not call state modifying function 'some_side_effects' within a range expression.`
I could not find a consistent way to test the side effect, but I've added the caching in [Cache the range end when complex](https://github.com/vyperlang/vyper/pull/3679/commits/4aa953f16a1137f0bfb3001f7bec51f43b4689b1)

Reply:
in the new IR it's no longer necessary anyways -- but for now, it's at least good hygiene.